hyperspace/parachain/src/chain.rs

```diff
@@ -180,4 +198,163 @@ where
+				Some(block) => {
+					log::info!("block query took {}", now.elapsed().as_millis());
+					break block
+				},
+				None => {
+					if now.elapsed() > Duration::from_secs(20) {
+						return Err(primitives::error::Error::from(
+							"Timeout while waiting for block".to_owned(),
+						))
+					}
+					sleep(Duration::from_millis(100)).await;
+				},
+			}
+		};
+		let extrinsic_opaque =
+			block.block.extrinsics.get(transaction_index).expect("Extrinsic not found");
+
+		let unchecked_extrinsic = UncheckedExtrinsic::<T>::decode(&mut &*extrinsic_opaque.encode())
+			.map_err(|e| {
+				primitives::error::Error::from(format!("Extrinsic decode error: {}", e))
+			})?;
+
+		match unchecked_extrinsic.function {
+			RuntimeCall::Ibc(IbcCall::deliver { messages }) => {
+				let message = messages.get(event_index).ok_or_else(|| {
+					primitives::error::Error::from(format!(
+						"Message index {} out of bounds",
+						event_index
+					))
+				})?;
+				let envelope = Ics26Envelope::<LocalClientTypes>::try_from(Any {
+					type_url: String::from_utf8(message.type_url.clone())?,
+					value: message.value.clone(),
+				});
+				match envelope {
+					Ok(Ics26Envelope::Ics2Msg(ClientMsg::UpdateClient(update_msg))) =>
+						return Ok(update_msg.client_message),
+					_ => (),
+				}
+			},
+			_ => (),
+		}
+		Err(primitives::error::Error::Custom("No ICS02 update message found".into()))
+	}
+}
+
+#[async_trait::async_trait]
+impl<T: Config + Send + Sync> MisbehaviourHandler for ParachainClient<T>
+where
+	u32: From<<<T as Config>::Header as HeaderT>::Number>,
+	u32: From<<T as Config>::BlockNumber>,
+	<T::Signature as Verify>::Signer: From<MultiSigner> + IdentifyAccount<AccountId = T::AccountId>,
+	MultiSigner: From<MultiSigner>,
+	<T as subxt::Config>::Address: From<<T as subxt::Config>::AccountId>,
+	T::Signature: From<MultiSignature>,
+	T::BlockNumber: BlockNumberOps + From<u32> + Display + Ord + sp_runtime::traits::Zero + One,
+	T::Hash: From<sp_core::H256> + From<[u8; 32]>,
+	FinalityProof<sp_runtime::generic::Header<u32, sp_runtime::traits::BlakeTwo256>>:
+		From<FinalityProof<T::Header>>,
+	BTreeMap<sp_core::H256, ParachainHeaderProofs>:
+		From<BTreeMap<<T as subxt::Config>::Hash, ParachainHeaderProofs>>,
+	sp_core::H256: From<T::Hash>,
+	<T::ExtrinsicParams as ExtrinsicParams<T::Index, T::Hash>>::OtherParams:
+		From<BaseExtrinsicParamsBuilder<T, AssetTip>> + Send + Sync,
+{
+	async fn check_for_misbehaviour<C: Chain>(
+		&self,
+		counterparty: &C,
+		client_message: AnyClientMessage,
+	) -> Result<(), anyhow::Error> {
+		match client_message {
+			AnyClientMessage::Grandpa(ClientMessage::Header(header)) => {
+				let target_block_number = header
+					.finality_proof
+					.unknown_headers
+					.iter()
+					.max_by_key(|h| h.number)
+					.expect("unknown_headers always contain at least one header; qed")
+					.number;
+				let finalized_block_number =
+					*self.relay_client.rpc().block(None).await?.unwrap().block.header.number();
```

Comment:
```suggestion
					*self.relay_client.rpc().block(None).await?.unwrap().block.header.number();
```
We should handle this error instead of unwrapping.

Reply:
Isn't there always a block available?

Reply:
Ideally there should be.